### PR TITLE
Handle trace id as header for event driven apps

### PR DESCRIPTION
**Pull request by rahulPalliyalil**
### What

The headers are key-value pairs that are transparently passed
by Kafka between producers and consumers.By default the traceid predefined header will be added to every kafka producer message.There is also the option to add custom headers to kafka by doing the following

```go
	// Create Producer with channels and config
pChannels := kafka.CreateProducerChannels()
pConfig := &kafka.ProducerConfig{MaxMessageBytes: &cfg.KafkaMaxBytes}
producer, err := kafka.NewProducer(ctx, cfg.Brokers, cfg.ProducedTopic, pChannels, pConfig)
producer.AddHeader(key, value)
```
The consumers can then retrieve these headers by the `GetHeader` api as follows.

```go
// consumer loop
func consume(upstream chan kafka.Message) {
	for {
		msg := <-upstream
		value := msg.GetHeader(key)
		msg.Commit()
	}
}
```
Please note: This pr is now raised against kafka version 2 , as there are still consumers for this version. once this is approved and merged will create a similar pr for latest version 3.

### How to review

Please check if the changes make sense. 

### Who can review

Anyone except me.

### CODE REVIEW

File: README.md [hunk 1]
```diff
@@ -126,6 +126,31 @@ The `Closer` channel is used to signal to all the loops that they need to exit b
 
 After successfully closing a producer or consumer, the corresponding `Closed` channel is closed.
 
+### Headers
+
+The headers are key-value pairs that are transparently passed
+by Kafka between producers and consumers.By default the traceid predefined header will be added to veery kafka producer message.There is also the option to add custom headers to kafka by doing the following
```
Comment (nshumoogum) [inline README.md:132]: Typo

```suggestion
by Kafka between producers and consumers.By default the traceid predefined header will be added to every kafka producer message.There is also the option to add custom headers to kafka by doing the following
```

File: README.md [hunk 2]
```diff
@@ -126,6 +126,31 @@ The `Closer` channel is used to signal to all the loops that they need to exit b
...
+
+```go
+	// Create Producer with channels and config
```
Comment (nshumoogum) [inline README.md:135]: Unnecessary whitespace

```suggestion
// Create Producer with channels and config
```

File: README.md [hunk 3]
```diff
@@ -126,6 +126,31 @@ The `Closer` channel is used to signal to all the loops that they need to exit b
...
+producer.AddHeader(key, value)
+```
+The consumers can then retrieve these headers by the `GetHeader` api as follows.
+
+```go
+// consumer loop
+func consume(upstream chan kafka.Message) {
+	for {
+		msg := <-upstream
+		value := msg.GetHeader(key)
```
Comment (nshumoogum) [inline README.md:148]: So for the retrieving and setting traceID into context, I was hoping the application could do something like:

```
		ctx := msg.Context() // like how we do it for http requests to our apps
```

Comment (nshumoogum) [reply README.md:148]: Happy to have a conversation about this as I dont know if this is possible. If it is then the key in context needs to match what has been done in dp-net, so I think this is `request_id` but you will want to check this.

**Review by nshumoogum — CHANGES_REQUESTED** [review]

File: README.md (hunk 1 above)
Comment (rahulPalliyalil) [reply README.md]: done

File: README.md (hunk 2 above)
Comment (rahulPalliyalil) [reply README.md]: done.

File: README.md (hunk 3 above)
Comment (rahulPalliyalil) [reply README.md:148]: done.
